src/DynamoCore/DSEngine/AstBuilder.cs

```diff
@@ -83,6 +93,80 @@ public static IEnumerable<NodeModel> TopologicalSort(IEnumerable<NodeModel> node
             return sortedNodes.Where(nodeModels.Contains);
         }
 
+        /// <summary>
+        /// BFS and post-order traverse to ensure all inputs are visited in
+        /// order and before downstream node.
+        /// </summary>
+        /// <param name="node"></param>
+        /// <param name="nodeFlags"></param>
+        /// <param name="sortedNodes"></param>
+        private static void BFSTraverse(
+            NodeModel node, 
+            Dictionary<NodeModel, MarkFlag> nodeFlags, 
+            Queue<NodeModel> sortedNodes)
+        {
+            MarkFlag flag;
+            if (!nodeFlags.TryGetValue(node, out flag))
+            {
+                flag = MarkFlag.NoMark;
+                nodeFlags[node] = flag;
+            }
+
+            if (flag != MarkFlag.NoMark)
+                return;
+
+            nodeFlags[node] = MarkFlag.TempMark;
+
+            for (int i = 0; i < node.InPortData.Count; ++i)
+            {
+                Tuple<int, NodeModel> t;
+                if (!node.TryGetInput(i, out t))
+                    continue;
+
+                BFSTraverse(t.Item2, nodeFlags, sortedNodes);
+            }
+
+            sortedNodes.Enqueue(node);
+            nodeFlags[node] = MarkFlag.Marked;
+        }
+
+        /// <summary>
+        /// Topological sort *the whole graph*. If the input nodes are part of a
+        /// graph, it does not promise to generate a good topological order. For 
+        /// example, for the following graph:
+        /// 
+        ///         +---+
+        ///         | A | -----+
+        ///         +---+      +----> +---+
+        ///                           | C |
+        ///                    +----> +---+
+        ///         +---+      |
+        ///         | B | -----+
+        ///         +---+
+        /// 
+        /// Their ideal topological order is A -> B -> C. If the input is only {A, B}, 
+        /// it may return {B, A} which is not OK in node to code.
+        ///
+        /// Note it is much slower thanopologiclSort().
+        /// </summary>
+        /// <param name="nodes"></param>
+        /// <returns></returns>
+        public static IEnumerable<NodeModel> TopologicalSortForGraph(IEnumerable<NodeModel> nodes)
+        {
+            var nodeFlags = nodes.ToDictionary(node => node, _ => MarkFlag.NoMark);
+            var sortedNodes = new Queue<NodeModel>();
+
+            // Get roots of these nodes
+            var roots = nodes.Where(n => !n.OutputNodes.Any());
+            foreach (NodeModel candidate in roots)
+                BFSTraverse(candidate, nodeFlags, sortedNodes);
+
+            foreach (NodeModel candidate in SortCandidates(nodeFlags))
+                BFSTraverse(candidate, nodeFlags, sortedNodes);
+
+            return sortedNodes;
+        }
+
         private static IEnumerable<NodeModel> SortCandidates(Dictionary<NodeModel, MarkFlag> nodeFlags)
```

Comment:
Why is this method not named `FindUnvisitedNodes`? :cold_sweat:

Also, what if all nodes are visited, does `FirstOrDefault` below returns `null` and crashes things?